Support View usage for PIVOT operator

RIC06X opened the pull request:
### Description
Support create/select/drop view on stmt with pivot operator. 

Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/418
### Issues Resolved

Task: BABEL-4673

### Test Scenarios Covered ###
* **Use case based -**
  - VIEW OF 2 COLUMN SRC TABLE PIVOT
  - VIEW OF 3 COLUMN SRC TABLE PIVOT
  - VIEW OF 4 COLUMN SRC TABLE PIVOT
  - VIEW OF TOP PIVOT
  - VIEW OF DISTINCT PIVOT
  - VIEW OF UNION PIVOT
  - VIEW OF PIVOT IN SUBQUERY
  - VIEW OF PIVOT FUNCTION
  - VIEW OF PIVOT FUNCTION
  - VIEW OF CTE PIVOT AS SRC TABLE
  - VIEW OF CTE PIVOT AS SRC TABLE JOIN ANOTHER PIVOT
  - VIEW OF CTE TABLE AS PIVOT SRC TABLE
  - VIEW OF JOIN PIVOT
  - VIEW OF INNER JOIN PIVOT
  - VIEW OF LEFT JOIN PIVOT
  - VIEW OF RIGHT JOIN PIVOT
  - VIEW OF FULL JOIN PIVOT
  - VIEW OF CROSS JOIN PIVOT
  - VIEW OF COMMA JOIN PIVOT
  - VIEW OF COMMA CROSS JOIN PIVOT

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

RIC06X committed:
Support View usage for PIVOT operator

Support create/select/drop view on stmt with pivot operator

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X commented:
15.8 related upgrade tests will get passed after BABEL3_X_DEV commit merged

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4455,32 +4457,79 @@ bbf_pivot(PG_FUNCTION_ARGS)
+	 * if view_uuid is not defined (view_uuid is empty string), then we will get the pivot metadata from
+	 * FunctionCallInfo(fcinfo).
+	 */
+	if (strlen(view_uuid) == 0)
```
can this be NULL? safer to do `if (!view_uuid || strlen(view_uuid))`

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4455,32 +4457,79 @@ bbf_pivot(PG_FUNCTION_ARGS)
+	 */
+	if (strlen(view_uuid) == 0)
+	{	
+		if (fcinfo->context == NULL || !IsA(fcinfo->context, List) || list_length((List *) fcinfo->context) != 3)
```
nit: I am not a huge fan of hardcoding values, in this case `3`. You can define a constant with descriptive name and use that instead.

RIC06X replied:
defined a constant called `PIVOT_METADATA_COUNT`

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4455,32 +4457,79 @@ bbf_pivot(PG_FUNCTION_ARGS)
+			BTEqualStrategyNumber, F_TEXTEQ,
+			CStringGetTextDatum(view_uuid));
+		
+		scan = systable_beginscan(rel, get_bbf_pivot_view_idx_oid(),
```
Dont you need to end the scan?

RIC06X replied:
ended scan in latest revision

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4455,32 +4457,79 @@ bbf_pivot(PG_FUNCTION_ARGS)
+		SysScanDesc	scan;
+		ScanKeyData	scanKey[1];
+
+		rel = table_open(get_bbf_pivot_view_oid(), RowExclusiveLock);
```
do you need to close the relation?

RIC06X replied:
closed in latest revision

lejaokri commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2751,6 +2803,60 @@ get_partition_depend_table_oid(HeapTuple tuple, TupleDesc dsc)
+static Datum
+get_pivot_view_schema_name(HeapTuple tuple, TupleDesc dsc)
+{
+	bool		schema_is_null, dbid_is_null;
+	char		*physical_schema_name, *schema_name, *org_schema_name;
```
nit: it's good practice to initialize variables and not trust the functions such as `heap_getattr()` to initialize them.

RIC06X replied:
all vars are initialized to `NULL` in latest revision

RIC06X commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -387,6 +387,20 @@ extern Oid	bbf_extended_properties_idx_oid;
 extern Oid	get_bbf_extended_properties_oid(void);
 extern Oid	get_bbf_extended_properties_idx_oid(void);
 
+typedef struct FormData_bbf_extended_properties
```
reorder this struct definition as it was moved away from its function declaration by previous commits

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4648,8 +4729,30 @@ get_bbf_pivot_tuplestore(RawStmt 	*sql,
+	 * if view_uuid is not defined, then we will use the rawparsetree of source_sql to get the
+	 * result
+	 */
+	if (!view_uuid || strlen(view_uuid) != 0)
```
shouldn't this be `if (view_uuid && strlen(view_uuid) != 0)`?

RIC06X replied:
my dumb mistake again 😥

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4559,8 +4617,30 @@ load_categories_hash(RawStmt *cats_sql, const char * sourcetext, MemoryContext p
+	 * if view_uuid is not defined, then we will use the rawparsetree of catefory_sql to get the
+	 * result
+	 */
+	if (!view_uuid || strlen(view_uuid) != 0)
```
same here. `if (!view_uuid || strlen(view_uuid) != 0)` seems wrong to me

RIC06X replied:
my dumb mistake😥

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4648,8 +4729,30 @@ get_bbf_pivot_tuplestore(RawStmt 	*sql,
+		initStringInfo(&buf);
+		appendStringInfoString(&buf, "select * from ");
+		appendStringInfoString(&buf, BBF_PIVOT_VIEW_SRC_PREFIX);
+		appendStringInfoString(&buf, view_uuid);
```
nit: where is `view_uuid` value coming from? is  it possible for it to have space in the beginning?

RIC06X replied:
catalog.c:6122

RIC06X replied:
no space is generated in the beginning

lejaokri commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5089,28 +5102,66 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+		char		*view_name;
+		char		*view_uuid;
+		view_name = getQueryEnvViewName(pstate->p_queryEnv);
+		view_uuid = add_entry_to_bbf_pivot_view(view_name,
```
are we allowed to use PIVOT in Read-only node? If yes, then this wouldnt work because it does `CatalogInsert`

lejaokri replied:
This will also have a side-effect on reader node such that a SELECT + pivot now allocates transaction id. Is this expected and discussed during design reviews?

RIC06X replied:
Yes, we can use it. Case 1 only applies to create pivot view. so select pivot will not fall into this scenario. So it will not have side effect on reader.

lejaokri replied:
thanks for clarifying!

RIC06X committed:
Add UUID check and validate function

Added pivot view UUID check and validate function to make sure UUID is
legit

Task:BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
Simplified bbf_pivot execution logic

Add pivot view metadata pre-execution logic to
pass_pivot_data_to_fcinfo. Simplified bbf_pivot execution logic to only
take and execute rawparsetree metadata.

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

rishabhtanwar29 commented on contrib/babelfishpg_tsql/sql/object_definition_tsql.sql:
```diff
@@ -11,4 +11,15 @@ CREATE TABLE sys.babelfish_view_def (
+);
+
+SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_view_def', '');
+SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_pivot_view', '');
```
Nit: Let's add a newline at the end of file.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.2.0--4.3.0.sql:
```diff
@@ -188,11 +188,27 @@ CREATE TABLE sys.babelfish_partition_depend
+CREATE OR REPLACE FUNCTION sys.bbf_pivot(IN arg TEXT)
+RETURNS setof record
+AS 'babelfishpg_tsql', 'bbf_pivot'
+LANGUAGE C STABLE;
```
I think this will create a new overloaded function instead of replacing the existing one. We would need to drop the previous definition and then re-create it if we intend to redefine the function.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -5908,6 +6030,290 @@ clean_up_bbf_partition_metadata(int16 dbid)
+
+
+static ViewStmt *
+make_pivot_view_helper_view(Node* node, const char *viewName, const char *schemaname)
```
We should rather say `makeViewStmt`

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -5908,6 +6030,290 @@ clean_up_bbf_partition_metadata(int16 dbid)
+ * 		Add a new entry to the sys.bbf_pivot_view catalog table.
+ */
+char *
+add_entry_to_bbf_pivot_view(const char *pivot_view_name, Node *src_parsetree, Node *cat_parsetree, const char * source_text, const char *agg_func_name)
```
Nit: wrap long line into multiple short lines.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -5908,6 +6030,290 @@ clean_up_bbf_partition_metadata(int16 dbid)
+	StringInfoData 		buf;
+
+	logical_schema_name = get_authid_user_ext_schema_name(get_cur_db_name(), get_user_for_database(get_cur_db_name()));
+	physical_schema_name = get_physical_schema_name(get_cur_db_name(), logical_schema_name);
```
Why are fetching the default schema name instead of what was specified in CREATE VIEW statement? This way we will end creating helper views in default schema where as the actual view might exists in a different schema (user specified) as well as we will be storing default schema name instead of actual schema in the catalog.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -5908,6 +6030,290 @@ clean_up_bbf_partition_metadata(int16 dbid)
+	DropStmt *n;
+
+	n = makeNode(DropStmt);
+	n->objects = list_make1(list_make1(makeString(view_name)));
```
We should schema qualify the object just to be safe.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -5908,6 +6030,290 @@ clean_up_bbf_partition_metadata(int16 dbid)
+    tuple = systable_getnext(scan);
+	while (HeapTupleIsValid(tuple))
+	{
+		if (HeapTupleIsValid(tuple))
```
Why checking tuple validity twice?

RIC06X committed:
Fixed default schema is always used when creating helper views

Task: BABEL-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
Refactor code and used deparsing approach for pivot view

Updated pivot view approach to deparsing approach. This approach will
deparse rawStmt to psql-text at analyzer stage which will be set in
bbf_pivot function argument. This eliminated the need for 2 new helper
view and an additional babel catalog.

Task: Babel-4673
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4685,18 +4688,19 @@ bbf_pivot(PG_FUNCTION_ARGS)
 	bbf_pivot_cat_sql = (RawStmt *) list_nth(pivot_parsetree, 1);
-	query_string = ((String *) list_nth(pivot_extrainfo, 0))->sval;
-	funcName = ((String *) list_nth(pivot_extrainfo, 1))->sval;
+	s_sql_string = text_to_cstring(PG_GETARG_TEXT_PP(0));
```
wil update s_sql_string to src_sql_string and c_sql_string to cat_sql_string

RIC06X commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4775,7 +4782,9 @@ load_categories_hash(RawStmt *cats_sql, const char * sourcetext, MemoryContext p
 		elog(ERROR, "load_categories_hash: SPI_connect returned %d", ret);
 
-	/* Retrieve the category name rows */
+	/* 
```
will update comment to 1 line

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5232,39 +5258,147 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+
+		src_sql_string = pstrdup(pstate->p_sourcetext);
+		cat_sql_string = pstrdup(pstate->p_sourcetext);
+
```
will remove the white space here

RIC06X commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -5232,39 +5258,147 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+		cat_sql_string =  text_to_cstring(PG_GETARG_TEXT_PP(1));
+
+		restored_src_rawstmt = linitial(raw_parser(src_sql_string, RAW_PARSE_DEFAULT));
+		restored_cat_rawstmt = linitial(raw_parser(cat_sql_string, RAW_PARSE_DEFAULT));
```
we don't have a parsing error here probably because we skipped the ANTLR parser. To make it more protected, I will guard those two with set sql dialect to psql and then tsql.
